processor/src/trace/mod.rs

```diff
@@ -82,29 +82,38 @@ impl ExecutionTrace {
         // perfect zero knowledge.
         let program_hash: Digest = process.decoder.program_hash().into();
         let rng = RandomCoin::new(&program_hash.to_bytes());
+
+        // create a new program info instance with the underlying kernel
+        let kernel = process.kernel.clone();
+        let program_info = ProgramInfo::new(program_hash, kernel);
         let (main_trace, aux_trace_hints) = finalize_trace(process, rng);
 
         Self {
             meta: Vec::new(),
             layout: TraceLayout::new(TRACE_WIDTH, [AUX_TRACE_WIDTH], [AUX_TRACE_RAND_ELEMENTS]),
             main_trace: Matrix::new(main_trace),
             aux_trace_hints,
-            program_hash,
-            program_outputs,
+            program_info,
+            stack_outputs,
         }
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
+    /// Returns the program info of this execution trace.
+    pub fn program_info(&self) -> &ProgramInfo {
+        &self.program_info
+    }
+
     /// Returns hash of the program execution of which resulted in this execution trace.
-    pub fn program_hash(&self) -> Digest {
-        self.program_hash
+    pub fn program_hash(&self) -> &Digest {
+        self.program_info.program_hash()
     }
```

Comment:
Since we have a new method for `program_info()`, I'd probably get rid of this one.

Reply:
I think using chained getters increases the coupling between the consumers of the current structure and its internal implementations.

Whenever I face a `foo.bar().baz()`, I prefer to create a `foo.baz()` that will internally call `foo.bar.baz()` so if I ever move `baz` to `foo`, the consumers of `foo` won't have to update their code as they will continue to call just `foo.baz()`. wdyt?